File: docs/chapters/01-solid-and-core-engineering-principles/go/open_closed_fees.go
// Scenario: checkout fee rules for a payments service (flat, percent, regional)
// Demonstrates: Open/Closed — register new FeeRule without editing Calculator
// Trade-off: map registry vs switch; Go uses small interfaces + registration

package main

import (
	"fmt"
	"math/big"
)

type FeeRule interface {
	ID() string
	Apply(subtotal *big.Rat, ctx map[string]string) *big.Rat
}

type flatFee struct{ id string; amount *big.Rat }

func (f flatFee) ID() string { return f.id }
func (f flatFee) Apply(_ *big.Rat, _ map[string]string) *big.Rat { return new(big.Rat).Set(f.amount) }

type FeeCalculator struct{ rules map[string]FeeRule }

func NewFeeCalculator(rules []FeeRule) *FeeCalculator {
	m := make(map[string]FeeRule, len(rules))
	for _, r := range rules {
		m[r.ID()] = r
	}
	return &FeeCalculator{rules: m}
}

func (c *FeeCalculator) Total(ruleID string, subtotal *big.Rat, ctx map[string]string) (*big.Rat, error) {
	r, ok := c.rules[ruleID]
	if !ok {
		return nil, fmt.Errorf("unknown rule: %s", ruleID)
	}
	return r.Apply(subtotal, ctx), nil
}

File: docs/chapters/01-solid-and-core-engineering-principles/java/OpenClosedFees.java
// Scenario: checkout fee rules for a payments service (flat, percent, regional)
// Demonstrates: Open/Closed — register new FeeRule without editing Calculator
// Trade-off: small registry vs giant switch; YAGNI if only one rule exists

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

interface FeeRule {
    String id();
    BigDecimal apply(BigDecimal subtotal, Map<String, String> ctx);
}

final class FlatFeeRule implements FeeRule {
    private final String id;
    private final BigDecimal amount;
    FlatFeeRule(String id, BigDecimal amount) { this.id = id; this.amount = amount; }
    public String id() { return id; }
    public BigDecimal apply(BigDecimal subtotal, Map<String, String> ctx) { return amount; }
}

final class FeeCalculator {
    private final Map<String, FeeRule> rules;
    FeeCalculator(List<FeeRule> rules) {
        this.rules = rules.stream().collect(
            java.util.stream.Collectors.toMap(FeeRule::id, Function.identity()));
    }
    BigDecimal total(String ruleId, BigDecimal subtotal, Map<String, String> ctx) {
        FeeRule rule = rules.get(ruleId);
        if (rule == null) throw new IllegalArgumentException("unknown rule: " + ruleId);
        return rule.apply(subtotal, ctx);
    }
}
